domain/agentprovisioner/service/service.go

```diff
@@ -0,0 +1,262 @@
+		st:             st,
+	}
+}
+
+// ContainerManagerConfigForType returns the container manager config for the
+// given container type.
+func (s *Service) ContainerManagerConfigForType(
+	ctx context.Context,
+	containerType instance.ContainerType,
+) (containermanager.Config, error) {
+	rval := containermanager.Config{
+		ModelID: s.modelID,
+	}
+
+	cfg, err := s.st.GetModelConfigKeyValues(ctx,
+		config.LXDSnapChannel,
+		config.ContainerImageMetadataURLKey,
+		config.ContainerImageMetadataDefaultsDisabledKey,
+		config.ContainerImageStreamKey,
+	)
+	if err != nil {
+		return containermanager.Config{}, fmt.Errorf(
+			"cannot get model config keys when calculating container manager config: %w",
+			err,
+		)
+	}
+
+	networkingMethod, err := s.determineContainerNetworkingMethod(ctx, cfg[config.ContainerNetworkingMethod])
+	if err != nil {
+		return containermanager.Config{}, fmt.Errorf("cannot determine container networking method: %w", err)
+	}
+	rval.NetworkingMethod = networkingMethod
+
+	if containerType == instance.LXD {
+		rval.LXDSnapChannel = cfg[config.LXDSnapChannel]
+	}
+	rval.ImageMetadataURL = cfg[config.ContainerImageMetadataURLKey]
+	rval.MetadataDefaultsDisabled, _ = strconv.ParseBool(cfg[config.ContainerImageMetadataDefaultsDisabledKey])
+	rval.ImageStream = cfg[config.ContainerImageStreamKey]
+
+	return rval, nil
+}
+
+// determineContainerNetworkingMethod consults the passed-in provider and
+// specified model config value to determine the networking method for the
+// container.
+func (s *Service) determineContainerNetworkingMethod(
+	ctx context.Context,
+	userDefinedNetworkingMethod string,
+) (containermanager.NetworkingMethod, error) {
+	provider, err := s.providerGetter(ctx)
+	if errors.Is(err, errors.NotSupported) {
+		// Provider doesn't supports container addresses
+		return containermanager.NetworkingMethodLocal, nil
+	}
+	if err != nil {
+		return "", fmt.Errorf(
+			"cannot get networking provider for model: %w",
+			err,
+		)
+	}
+
+	method := containermanager.NetworkingMethod(userDefinedNetworkingMethod)
+	if method != containermanager.NetworkingMethodUndefined && method != containermanager.NetworkingMethodAuto {
+		// User specified a container networking method - use that
+		return method, nil
+	}
+
+	// Auto-configure container networking method
+	method = containermanager.NetworkingMethodLocal
+	if provider != nil {
+		supports, err := provider.SupportsContainerAddresses(envcontext.WithoutCredentialInvalidator(ctx))
+		if err != nil {
+			return "", fmt.Errorf(
+				"cannot determine if provider supports container addresses: %w",
+				err,
+			)
+		}
+		if supports {
+			return containermanager.NetworkingMethodProvider, nil
+		}
+	}
+	return method, nil
+}
+
+// ContainerConfig returns the container config for the model.
+func (s *Service) ContainerConfig(ctx context.Context) (params.ContainerConfig, error) {
+	result := params.ContainerConfig{}
+
+	modelConfig, err := s.st.GetModelConfigKeyValues(ctx,
```

Comment:
This does not look nice at all.

Reply:
no, I agree, it's disgusting :)

I could store the list of keys in a const/var somewhere?